runtime/compiler/trj9/optimizer/JProfilingRecompLoopTest.cpp

```diff
@@ -0,0 +1,214 @@
+      
+      TR::Block *originalBlock = asyncCheckTreeTop->getEnclosingBlock();
+      TR::Block *newSplittedBlock = originalBlock->split(asyncCheckTreeTop->getNextTreeTop(), cfg, true, true);
+      TR::Block *callRecompileBlock = TR::Block::createEmptyBlock(node, comp, UNKNOWN_COLD_BLOCK_COUNT);
+      callRecompileBlock->setIsCold(true);
+      
+      TR::Node *callNode = TR::Node::createWithSymRef(node, TR::icall, 2, comp->getSymRefTab()->findOrCreateRuntimeHelper(TR_jitRetranslateCallerWithPrep, false, false, true));
+      callNode->setAndIncChild(0, TR::Node::createWithSymRef(node, TR::loadaddr, 0, comp->getSymRefTab()->findOrCreateStartPCSymbolRef()));
+      callNode->setAndIncChild(1, TR::Node::createWithSymRef(node, TR::loadaddr, 0, comp->getSymRefTab()->findOrCreateCompiledMethodSymbolRef()));
+      TR::TreeTop *callTree = TR::TreeTop::create(comp, TR::Node::create(node, TR::treetop, 1, callNode));
+      callTree->getNode()->setIsProfilingCode();
+      callRecompileBlock->append(callTree);
+      cfg->addNode(callRecompileBlock);
+      
+      TR::Node *loadBaseThreshold = TR::Node::createWithSymRef(node, TR::iload, 0, comp->getSymRefTab()->createKnownStaticDataSymbolRef(&recompileThreshold, TR::Int32));
+      TR::Node *adjustThreshold = TR::Node::create(node, TR::ishl, 2, loadBaseThreshold, TR::Node::iconst(node, (depth-1)*3));
+      TR::Node *cmpNode = TR::Node::createif(TR::ificmple, root, adjustThreshold, newSplittedBlock->getEntry());
+      
+      TR::TreeTop *cmpFlag = TR::TreeTop::create(comp, asyncCheckTreeTop, cmpNode);
+      newSplittedBlock->getEntry()->insertTreeTopsBeforeMe(callRecompileBlock->getEntry(), callRecompileBlock->getExit());
+      cfg->addEdge(TR::CFGEdge::createEdge(callRecompileBlock, newSplittedBlock, comp->trMemory()));
+      cfg->addEdge(TR::CFGEdge::createEdge(originalBlock, callRecompileBlock, comp->trMemory()));
+      if (trace())
+         traceMsg(comp,"\t\t Newly created recompilation Test : n%dn in block_%d\n/\t\tRecompilation Call in block_%d\n",
+            callNode->getGlobalIndex(), originalBlock->getNumber(), callRecompileBlock->getNumber());
+      }
+   }
+
+TR::Node *
+TR_JProfilingRecompLoopTest::generateBlockFrequecnyCalculationNode(TR::Compilation *comp, TR::Node *node, TR_BlockFrequencyInfo *bfi, int32_t blockNumber)
+   {
+   TR_BitVector **componentCounters = bfi->getCounterDerivativeInfo();
+   TR::Node *root = NULL;
+   if (blockNumber > -1 && (componentCounters[blockNumber * 2]
+      && (((uintptr_t)componentCounters[blockNumber * 2]) & 0x1 == 1)
+      || !componentCounters[blockNumber * 2]->isEmpty()))
+      {
+      TR::Node *addRoot = NULL;
+      if (((uintptr_t)componentCounters[blockNumber * 2]) & 0x1 == 1)
+         {
+         TR::SymbolReference *symRef = comp->getSymRefTab()->createKnownStaticDataSymbolRef(bfi->getFrequencyForBlock(((uintptr_t)componentCounters[blockNumber * 2]) >> 1), TR::Int32);
```

Comment:
the logic for the static address should be commoned out since several people need that.